### ISSUE TITLE
Proposal: Steal F#'s nice union type syntax

### ISSUE THREAD

**Issue by rtfeldman** [elm-lang/elm-compiler#1385]
tl;dr write union types with a `|` before each case, like so:

```elm
type Maybe a =
    | Just a
    | Nothing
```

# Motivation

The current syntax for defining union types has these problems:

* It causes needless version control diffs when removing a tag that happens to be in the first position
* Its multi-line definitions are visually inconsistent with other definitions in Elm, which have `=` at the end of the line instead of the beginning of the following line
* Its multi-line definitions are visually inconsistent with its own single-line definitions, meaning switching from single-line to multi-line also entails a larger-than-necessary version control diff

# Proposed solution

Do [what F# does](https://fsharpforfunandprofit.com/posts/discriminated-unions/): require that each tag be on a different line, and preceded by a `|`.

# Examples

### Multi-Line

From [`evancz/elm-todomvc`](https://github.com/evancz/elm-todomvc/blob/425d1711dbd63815a2f3ea0d24497055d8b7c748/Todo.elm#L99-L109)

```elm
type Msg
    = NoOp
    | UpdateField String
    | EditingTask Int Bool
    | UpdateTask Int String
    | Add
    | Delete Int
    | DeleteComplete
    | Check Int Bool
    | CheckAll Bool
    | ChangeVisibility String
```

becomes

```elm
type Msg =
    | NoOp
    | UpdateField String
    | EditingTask Int Bool
    | UpdateTask Int String
    | Add
    | Delete Int
    | DeleteComplete
    | Check Int Bool
    | CheckAll Bool
    | ChangeVisibility String
```

### Single-Line

This is currently supported, but would no longer be.

```elm
type Bool = True | False
```

becomes 

```elm
type Bool =
    | True
    | False
```

### Split-Line

This is currently supported, but would no longer be.

```elm
type Maybe a = Just a
    | Nothing
```

becomes 

```elm
type Maybe a =
    | Just a
    | Nothing
```

**Comment by rtfeldman** [comment]
As an aside, I don't think this is in any way urgent. I just think it's a good idea and wanted to document it.

**Comment by mgold** [comment]
What if the preceding bar was optional, not required? That would be a non-breaking syntax change.

**Comment by rtfeldman** [comment]
I'd prefer that to status quo, since then I could write my stuff in the
nice way, so sure! :)

**Comment by pavelhritonenko** [comment]
In F# it's optional too

![image](https://cloud.githubusercontent.com/assets/2510026/15319699/ec7c7390-1c34-11e6-8f5a-10536c00ccaf.png)

+1

**Comment by evancz** [comment]
Thanks. Added to #1375. In case any OCaml people ever see this, they did it first ;)

**Comment by 7sharp9** [comment]
As a side note, with a single case union in F# the first bar is entirely optional:

So
```fsharp
type Msg = UpdateField of string
```
Is valid syntax rather than having to do:
```fsharp
type Msg =
    | UpdateField of string
```

**Comment by altaic** [comment]
If single line unions are no longer allowed as proposed, why not just get rid of the `|` syntax altogether? Just use indentation as with `case`.

**Comment by Heimdell** [comment]
Maybe, leading bar could be made optional?

**Comment by mgold** [comment]
@Heimdell See my May 16th comment...
